misc/python/materialize/cli/ci_coverage_pr_report.py

```diff
@@ -0,0 +1,90 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import re
+import subprocess
+import sys
+from collections import OrderedDict
+from itertools import groupby
+from typing import Dict, List
+
+Coverage = Dict[str, OrderedDict[int, int]]
+SOURCE_RE = re.compile(
+    "^/var/lib/buildkite-agent/builds/buildkite-.*/materialize/tests/(.*$)"
+)
+
+
+def find_interesting_lines() -> Coverage:
+    coverage: Coverage = {}
+    # TODO: Get base state from buildkite instead of just HEAD~
+    result = subprocess.run(
+        ["git", "diff", "-U0", "HEAD~"], check=True, capture_output=True
+    )
+    file = None
+    for line_raw in result.stdout.splitlines():
+        line = line_raw.decode("utf-8")
+        if line.startswith("+++"):
+            if not line.endswith(".rs"):
+                continue
+            file = line.removeprefix("+++ b/")
+            coverage[file] = OrderedDict()
+        elif line.startswith("@@ ") and file:
+            parts = line.split(" ")[2]
+            if "," in parts:
+                start, length = map(int, parts.split(","))
+            else:
+                start = int(parts)
+                length = 1
+            for line_nr in range(start, start + length):
+                coverage[file][line_nr] = 0
+    return coverage
+
+
+def mark_covered_lines(lcov_file: str, coverage: Coverage) -> None:
+    for line in open(lcov_file):
+        line = line.strip()
+        if not line:
+            continue
+        if line == "end_of_record":
+            continue
+        method, content = tuple(line.strip().split(":", 1))
+        if method == "SF":
+            result = SOURCE_RE.search(content)
+            assert result
+            file = result.group(1)
+        elif method == "DA" and file in coverage:
+            line_str, hit_str = content.split(",", 1)
+            line_nr = int(line_str)
+            hit = int(hit_str) if hit_str.isnumeric() else int(float(hit_str))
+            if line_nr in coverage[file]:
+                coverage[file][line_nr] += hit
+
+
+def print_report(coverage: Coverage) -> None:
+    print("~~~ Uncovered Lines Changed in Pull Request")
+    for file, lines in coverage.items():
+        for key, group_iter in groupby(enumerate(lines), key=lambda x: x[1] - x[0]):
+            group = list(group_iter)
+            start = group[0][1]
+            end = group[-1][1]
+            if start == end:
```

Comment:
I think this needs to print a diff, otherwise the tool will be difficult to use for all parties involved, which will impact its adoption.

The easiest way to get to a diff is to simply delete the uncovered lines directly from the checked-out source, so that `git diff` will give a readable report directly, with the uncovered lines listed as `-`. Or, in other words, open the source file and write it back in-place, omitting any uncovered lines as you write them out.

Reply:
Great idea, thanks

Reply:
My proper solution would be to mark this inline in github as comments, but I'll leave that as an extension for later.